cmd: Add `caddy fmt` command.

vrongmeal opened the pull request:
This takes the config file as input and formats it.

Code currently implements a `Formatter` interface
which can be implemented by various adapter to format
a file.

Currently only Caddyfile formatter is implemnented.

Fixes #3020 

I still have to write tests (`TODO`).

An example of formatting (I hope this covers everything):

**Before `caddy format`:**

```
# Caddyfile with { comments } and "quotes"
# and improper \indentation

localhost { # starts with a comment
{$SOME_ENV}
		encode zstd gzip
  templates
 file_server browse\this\server
}

:8080, :8081 { respond "I am \"80\""
}
```

**After `caddy format`:**

```
# Caddyfile with { comments } and "quotes"
# and improper \indentation

localhost {
	# starts with a comment
	{$SOME_ENV}
	encode zstd gzip
	templates
	file_server browse\this\server
}

:8080, :8081 {
	respond "I am \"80\""
}
```

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/caddyserver/caddy?pullRequest=3090) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/caddyserver/caddy?pullRequest=3090) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/caddyserver/caddy?pullRequest=3090) it.</sub>

francislavoie commented:
Awesome! Could we call it `caddy fmt`? The idea was to match [`gofmt`](https://golang.org/cmd/gofmt/) 😅

francislavoie commented:
Awesome! Could we call it `caddy fmt`? The idea was to match [`gofmt`](https://golang.org/cmd/gofmt/) 😅

vrongmeal commented:
> Awesome! Could we call it caddy fmt? The idea was to match gofmt 😅

Yeah! I'll change that. Also right now it takes a flag `--config` as the name of the file (just like `caddy adapt` and `caddy validate`). I guess we could change that to an argument as well.

mholt commented on cmd/commandfuncs.go:
```diff
@@ -538,6 +538,37 @@ func cmdValidateConfig(fl Flags) (int, error) {
+	output := formatter.Format(input)
+	// fmt.Print(string(output))
+
+	err = ioutil.WriteFile(formatCmdConfigFlag, output, 0644)
```
I think the command must not mutate the input file. Dangerous and unexpected! Print to stdout instead. User can redirect to a file if they want. Or we can add a flag for it. This would be my first suggestion for a revision.

francislavoie replied:
FWIW I'd actually expect it to mutate the input. That's how most `fmt` things work. I feel like having a `--dry-run` option or similar would make more sense, which would just output to stdout.

mholt replied:
@francislavoie 

>  That's how most fmt things work.

Citation needed. 😁  (But they won't change my mind, I'm afraid.)

Here, I'll give you a counterexample: https://golang.org/cmd/gofmt/ -- does not overwrite the input.

Overwriting input files is risky, and I feel pretty strongly that we shouldn't do that.

If you want to do that, it's easy with my proposed change:

```
caddy fmt > Caddyfile
```

This also makes the command more convenient for scripting and other use.

Plus, I've found that it's quite common in practice that the caddy process does not have write privileges on its input config file.

The sweet spot for this feature is potential editor integration in the future.

francislavoie replied:
> Here, I'll give you a counterexample: golang.org/cmd/gofmt -- does not overwrite the input.

Could've sworn it did. Haven't played with Go in a while so 🤷‍♂ I stand corrected

I don't feel super strongly about this point, so I concede 😄

tobya replied:
I'm with Matt here

vrongmeal replied:
> caddy fmt > Caddyfile

So I think I should also default `--config` to take `Caddyfile` so we can use it this way.

And is it better to accept file as an argument or flag?

roblabla replied:
FWIW `caddy fmt > Caddyfile` won't work because it will cause Caddyfile to be replaced with an empty file before `caddy fmt` runs. You can try it easily, the following will empty up file:

```
echo test > file
cat file > file
```

So you'd have to do it in a two step process of `caddy fmt > Caddyfile.tmp && mv Caddyfile.tmp Caddyfile`. Not a big deal though.

mholt replied:
Wow, I hate bash.

vrongmeal replied:
Yeah, and so I added the `--write` option :D

mholt commented on cmd/commands.go:
```diff
@@ -242,6 +242,19 @@ provisioning stages.`,
 	})
 
+	RegisterCommand(Command{
+		Name: "format",
```
Let's call it `fmt` as an homage to `go fmt` :)

vrongmeal replied:
I actually named it `format` since all the other command names were expressive and this seemed like an exception among them. But I do like `fmt` too.

mholt replied:
Gotcha. I'm torn. I don't care either way :)

vrongmeal commented:
Updated this:

1. Changed the name from `format` to `fmt`.
1. Changed path of config from flag to argument.
1. Default path is `Caddyfile` (so is the corresponding formatter).
1. Added `--write` flag if changes are to be written to the file.

mholt commented on caddyconfig/caddyformatters.go:
```diff
@@ -0,0 +1,41 @@
+	return configFormatters[name]
+}
+
+var configFormatters = make(map[string]Formatter)
```
As cool as it would be to have formatters be extensible, I really can't think of a need for formatters beyond just the Caddyfile. I mean, I wouldn't want/expect Caddy to be able to format any other general-purpose config language like NGINX configs or YAML or any other config adapters. That'd be kinda out of scope, right?

If you agree, I think we can simplify things a lot. This file wouldn't be needed. No need to register formatters. No `--formatter` flag. Just call the `Format()` function from the command, since there's only the one.

vrongmeal replied:
Umm, we could remove it. The first thing I thought about this was we could extend it to JSON for starters easily but I get why we wouldn't like to do that.

mholt commented on cmd/commands.go:
```diff
@@ -242,6 +242,21 @@ provisioning stages.`,
+		Usage: "[--formatter <name>] [--write] [<path>]",
+		Short: "Formats a configuration file",
+		Long: `
+Formats the configuration file and prints the result on standard output.`,
```
This doc text will need some expansion before the final merge (take cues from some of the other commands).

vrongmeal replied:
Yeah! I'll add more to it

mholt commented on cmd/commandfuncs.go:
```diff
@@ -538,6 +538,47 @@ func cmdValidateConfig(fl Flags) (int, error) {
+	output := formatter.Format(input)
+
+	if formatCmdWriteFlag {
+		err = ioutil.WriteFile(formatCmdConfigFile, output, 0644)
```
We must be careful not to change a file's permissions. Be sure the original permissions are preserved.

vrongmeal replied:
This actually does not change permissions. If the file does not exist (which in this case we're sure it does) it just creates a file with these permissions.

From docs:
> WriteFile writes data to a file named by filename. If the file does not exist, WriteFile creates it with permissions perm (before umask); otherwise WriteFile truncates it before writing.

mholt replied:
Ah, I stand corrected. Thanks!

vrongmeal commented:
Update:
1. Remove `Formatter` interface (support only Caddyfile).
1. Improve description of command.

mholt commented on caddyconfig/caddyfile/formatter.go:
```diff
@@ -0,0 +1,141 @@
+// Formatter is a type that implements caddyconfig.Formatter.
+// This is used to format a Caddyfile.
+type Formatter struct {
+}
```
This seems outdated now. This type is not needed, probably!

mholt commented on caddyconfig/caddyfile/formatter.go:
```diff
@@ -0,0 +1,141 @@
+}
+
+// Format formats the body of Caddyfile.
+func (f Formatter) Format(body []byte) []byte {
```
This method can just be a function!
```suggestion
// Format formats a Caddyfile to conventional standards.
func Format(body []byte) []byte {
```

mholt commented on cmd/commands.go:
```diff
@@ -242,6 +242,24 @@ provisioning stages.`,
+		Short: "Formats a Caddyfile",
+		Long: `
+Formats the Caddyfile by adding proper indentation and spaces to improve
+human readability. It prints the output on the STDOUT by default.
```
```suggestion
human readability. It prints the result to stdout.
```

vrongmeal commented:
Updated this with above suggestions.

Added tests too :)
